src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -269,8 +269,12 @@ case class MatrixRead(
       if (dropVariants)
         OrderedRDD.empty[Annotation, Annotation, (Annotation, Iterable[Annotation])](hc.sc)
       else {
+        val rowTTBc = BroadcastTypeTree(hc.sc, typ.rowType)
         val rdd = new ReadRowsRDD(hc.sc, path, typ.rowType, nPartitions)
-        .map { ur =>
+        .map { rv =>
+          // FIXME expensive
+          val ur = new UnsafeRow(rowTTBc, rv.region.copy(), rv.offset)
```

Comment:
The solution is to use a single unsafe row, `mapPartitions` and mutate the row?

Reply:
Solution is to never make `UnsafeRow`s.  The FIXME isn't really necessary, we know where we're going.

Yes, mapPartitions should make a single region which is mutated and we insert judicious copies only where necessary (in aggregators like collect, takeBy), VSM.collect, etc.  I have a PR for that ready once all this stuff goes in.